spec/abci++/abci++_methods_002_draft.md

```diff
@@ -0,0 +1,849 @@
+### PrepareProposal
+
+#### Parameters and Types
+
+* **Request**:
+
+    | Name                    | Type                                        | Description                                                                                                      | Field Number |
+    |-------------------------|---------------------------------------------|------------------------------------------------------------------------------------------------------------------|--------------|
+    | hash                    | bytes                                       | The block header's hash of the block to propose. Present for convenience (can be derived from the block header). | 1            |
+    | header                  | [Header](../core/data_structures.md#header) | The header of the block to propose.                                                                              | 2            |
+    | tx                      | repeated bytes                              | Preliminary list of transactions that have been picked as part of the block to propose.                          | 3            |
+    | byzantine_validators    | repeated [Evidence](#evidence)              | List of evidence of validators that acted maliciously.                                                           | 4            |
+    | last_commit_info        | [LastCommitInfo](#lastcommitinfo)           | Info about the last commit, including the round, the validator list, and which ones signed the last block.       | 5            |
+
+>**TODO**: Add the changes needed in LastCommitInfo for vote extensions
+
+>**TODO**: DISCUSS: We need to make clear whether a proposer is also running the logic of a non-proposer node (in particular "ProcessProposal")
+From the App's perspective, they'll probably skip ProcessProposal
+
+* **Response**:
+
+    | Name                    | Type                                             | Description                                                                                 | Field Number |
+    |-------------------------|--------------------------------------------------|---------------------------------------------------------------------------------------------|--------------|
+    | modified_tx             | bool                                             | The Application sets it to true to denote it made changes to transactions                   | 1            |
+    | tx                      | repeated [TransactionRecord](#transactionrecord) | Possibly modified list of transactions that have been picked as part of the proposed block. | 2            |
+    | app_hash                | bytes                                            | The Merkle root hash of the application state.                                              | 3            |
+    | tx_result               | repeated [TxResult](#txresult)                   | List of structures containing the data resulting from executing the transactions            | 4            |
```

Comment:
Why do we have this separate to the tx field. Why not add it to the `TransactionRecord` struct?

Reply:
I guess you did this because it matches the response of `FinalizeBlock` although it might be easier to index if we pair the `tx` and its result together

Reply:
Also if we're still going to execute `FinalizeBlock` why do we need to return the `tx_result`'s now. (We don't do it in `PrepareProposal`)

Reply:
There are to points here I'd like to touch on:

1. The reason I kept fields `tx` and `tx_result` separate is because `tx` always makes sense, whereas `tx_result` only makes sense in same-block execution. So if I am building an App in next-block execution mode, I'd find it confusing that an internal part of the struct should be ignored depending on the mode.
2. We need to provide `tx_results` here, in same-block mode, because Tendermint will re-calculate the block's header just after this call, and before broadcasting the proposal message.

As for point 1 above, we can reshuffle the structs if you feel strongly about it.

Reply:
> We need to provide tx_results here, in same-block mode, because Tendermint will re-calculate the block's header just after this call, and before broadcasting the proposal message.

Ah right that makes sense. I personally think we should get rid of the txresultshash but let's keep it there for now. How does this impact finalize block. Do we still expect the application to emit the same events in that call. 

Also for the non-proposer, don't we require them to also emit events in `ProcessProposal` so that we can check that the hashes match or are they responsible for making sure the hashes match? Is it okay to ask the application to do that: "Hey, if you want state machine replication, do you mind just hashing your responses and then making sure it equals the header we just gave you?"

Reply:
> Ah right that makes sense. I personally think we should get rid of the txresultshash but let's keep it there for now. How does this impact finalize block. Do we still expect the application to emit the same events in that call.

This is the way I see it:

* in next-block mode: `tx_results` only make sense in `FinalizeBlock`
* in same-block mode: `tx_results` only make sense in `PrepareProposal`

The same is true for the neighboring parameters: `app_hash`, `validator_updates`, and `consensus_param_updates`.

I explain this in the "Usage" section of both `PrepareProposal` and `FinalizeBlock`, but there's maybe a way to make it clearer in the text.

> Also for the non-proposer, don't we require them to also emit events in `ProcessProposal` so that we can check that the hashes match or are they responsible for making sure the hashes match? Is it okay to ask the application to do that: "Hey, if you want state machine replication, do you mind just hashing your responses and then making sure it equals the header we just gave you?"

Hmmm, I think you got a point here. In the same way as *everyone* needs to provide `tx_results` at `FinalizeBlock` so that Tendermint can check the hashes (next-block mode); we should allow *everyone* to provide the `tx_results` at `Pr****Proposal` so that Tendermint can also check the hashes (same-block mode). So I think we are missing those params in `ResponseProcessProposal`. I'll add them. Thanks for spotting this :-)